File: apps/swarm/planner.py
import json
import litellm
from typing import List
from libs.core.config import get_model_config

class Planner:
    def __init__(self, model_name: str = None):
        self.model_config = get_model_config(model_name)
        name = self.model_config.get("model_name", "gpt-4o")
        provider = self.model_config.get("provider")

        if provider and provider != "openai":
             self.model = f"{provider}/{name}"
        else:
             self.model = name

    def plan(self, goal: str) -> List[str]:
        """
        Breaks a goal into a list of specific subtasks using LLM.
        Returns a list of strings, where each string is a subtask.
        """
        prompt = f"""
        You are a Senior Project Manager for a software engineering team.
        Your goal is to break down a high-level request into smaller, isolated, actionable coding tasks.
        Each task must be able to be executed by a single developer on a single branch.

        Goal: {goal}

        Return the result STRICTLY as a JSON list of strings. Do not include markdown formatting.
        Example: ["Create utils.py file", "Add unit tests for utils.py", "Update main.py to use utils"]
        """

        try:
            response = litellm.completion(
                model=self.model,
                messages=[{"role": "user", "content": prompt}],
                response_format={ "type": "json_object" } # Helps if model supports it, otherwise prompt instructions
            )
            content = response.choices[0].message.content.strip()

            # Remove markdown code blocks if present
            if content.startswith("```json"):
                content = content[7:]
            if content.startswith("```"):
                content = content[3:]
            if content.endswith("```"):
                content = content[:-3]

            try:
                # Some models might return {"tasks": [...]}, others just [...]
                data = json.loads(content)
                if isinstance(data, list):
                    return data
                elif isinstance(data, dict):
                     # Search for the first list value
                     for val in data.values():
                         if isinstance(val, list):
                             return val
                     return [goal] # Fallback
                else:
                    return [goal]
            except json.JSONDecodeError:
                # Fallback if JSON parsing fails
                return [goal]

        except Exception as e:
            print(f"Planner Error: {str(e)}")
            return [goal]

File: crates/core/src/planner.rs
use crate::Client;
use koval_common::Config;
use koval_protocol::openai::{ChatCompletionRequest, Message};
use serde_json::Value;
use anyhow::{anyhow, Result};

pub struct Planner {
    client: Client,
    config: Config,
}

impl Planner {
    pub fn new(config: Config, client: Client) -> Self {
        Self { client, config }
    }

    pub async fn plan(&self, goal: &str) -> Result<Vec<String>> {
        let system_prompt = r#"
You are a Senior Technical Project Manager.
Your goal is to break down a high-level user request into a list of specific, isolated, actionable coding tasks.
Each task must be clear enough for a junior developer to execute independently.

Return the result strictly as a JSON list of strings.
Do not include markdown formatting or explanation.
Example: ["Create utils.rs file", "Add unit tests for utils.rs", "Update main.rs"]
"#;

        let request = ChatCompletionRequest {
            model: self.config.model.clone(),
            messages: vec![
                Message::system(system_prompt),
                Message::user(goal),
            ],
            tools: None,
            stream: true, // We still use stream internally but call 'chat' which collects it
        };

        let response = self.client.chat(request).await?;
        let full_content = response.content.unwrap_or_default();

        // Clean up markdown code blocks if present
        let clean_content = full_content
            .trim()
            .trim_start_matches("```json")
            .trim_start_matches("```")
            .trim_end_matches("```")
            .trim();

        let parsed: Value = serde_json::from_str(clean_content)
            .map_err(|e| anyhow!("Failed to parse planner JSON: {}. Content: {}", e, clean_content))?;

        if let Some(arr) = parsed.as_array() {
            let tasks: Vec<String> = arr.iter()
                .filter_map(|v| v.as_str().map(|s| s.to_string()))
                .collect();
            
            if tasks.is_empty() {
                Ok(vec![goal.to_string()]) // Fallback
            } else {
                Ok(tasks)
            }
        } else if let Some(obj) = parsed.as_object() {
             // Handle {"tasks": [...]} case
             if let Some(val) = obj.values().find(|v| v.is_array()) {
                 let tasks: Vec<String> = val.as_array().unwrap().iter()
                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
                    .collect();
                 Ok(tasks)
             } else {
                 Ok(vec![goal.to_string()])
             }
        } else {
            Ok(vec![goal.to_string()])
        }
    }
}
